agent/php_user_instrument_hashmap.c

```diff
@@ -0,0 +1,345 @@
+
+  count = nr_count_buckets(hashmap);
+  for (i = 0; i < count; i++) {
+    nr_wraprecs_bucket_t* bucket = hashmap->buckets[i];
+    collisions = 0;
+
+    if (bucket) {
+      stats.buckets_used++;
+    }
+
+    while (bucket) {
+      nr_wraprecs_bucket_t* next = bucket->next;
+
+      collisions++;
+
+      nr_destroy_wraprecs_bucket(&bucket, hashmap->dtor_func);
+      bucket = next;
+    }
+
+    if (0 != collisions && collisions < stats.collisions_min) {
+      stats.collisions_min = collisions;
+    }
+    if (collisions > stats.collisions_max) {
+      stats.collisions_max = collisions;
+    }
+    if (collisions > 1) {
+      stats.buckets_with_collisions++;
+      stats.collisions_mean += collisions;
+    }
+  }
+
+  if (0 != stats.buckets_with_collisions) {
+    stats.collisions_mean /= stats.buckets_with_collisions;
+  }
+
+  nr_free(hashmap->buckets);
+  nr_realfree((void**)hashmap_ptr);
+
+  return stats;
+}
+
+static bool nr_zf_is_unnamed_closure(const zend_function* zf) {
+  if (9 != ZSTR_LEN(zf->op_array.function_name)) {
+    return false;
+  }
+  return 0 == memcmp("{closure}", ZSTR_VAL(zf->op_array.function_name), 9);
+}
+
+void nr_php_wraprec_hashmap_key_set(nr_php_wraprec_hashmap_key_t* key,
+                                    const zend_function* zf) {
+  key->lineno = nr_php_zend_function_lineno(zf);
+  key->scope_name = NULL;
+  key->function_name = NULL;
+  key->filename = NULL;
+
+  if (NULL != zf->op_array.function_name && !nr_zf_is_unnamed_closure(zf)) {
+    key->function_name = zf->op_array.function_name;
+    zend_string_addref(key->function_name);
+    if (zf->op_array.scope) {
+      key->scope_name = zf->op_array.scope->name;
+      zend_string_addref(key->scope_name);
+    }
+  } else {
+    if (NULL != zf->op_array.filename) {
+      key->filename = zf->op_array.filename;
+      zend_string_addref(key->filename);
+    }
+  }
+}
+
+void nr_php_wraprec_hashmap_key_release(nr_php_wraprec_hashmap_key_t* key) {
+  if (NULL != key->scope_name) {
+    zend_string_release(key->scope_name);
+  }
+  if (NULL != key->function_name) {
+    zend_string_release(key->function_name);
+  }
+  if (NULL != key->filename) {
+    zend_string_release(key->filename);
+  }
+  key->lineno = 0;
+}
+
+static inline size_t nr_zendfunc2bucketidx(size_t log2_num_buckets,
+                                           zend_function* zf) {
+  /* default to lineno */
+  uint32_t hash = nr_php_zend_function_lineno(zf);
+
+  if (NULL != zf->op_array.function_name && !nr_zf_is_unnamed_closure(zf)) {
+    /* but use hash of function name when possible */
+    hash = ZSTR_HASH(zf->op_array.function_name);
+  } else if (NULL != zf->op_array.filename) {
+    /* and as last resort try hash of filename if available */
+    hash = ZSTR_HASH(zf->op_array.filename);
+  }
+
+  /* normalize to stay in-bound */
+  return (hash & ((1 << log2_num_buckets) - 1));
+}
+
+static inline bool zstr_equal(zend_string* zs1, zend_string* zs2) {
```

Comment:
Is this a function that we can extract into a common util file (like agent/php_zval.c)?

Reply:
Possibly. But this is not part of this PR.